### ISSUE TITLE
DRAFT: Replace Plone as a rest API with Nick

### ISSUE THREAD

**Issue by robgietema** [plone/Products.CMFPlone#4069]
## Abstract

The Plone content management system has been around for over 20 years and in that time a lot of functionality has been added and removed to the code base, which means a large codebase (~1.250.000 lines of code) needs to be maintained. Since a couple of years Plone moved to a headless solution using Volto as a frontend solution where Plone acts as an API service. Because of this a lot of functionality is not needed anymore in Plone.

Since switching to a JavaScript frontend a lot of new developer have come in who have no, or little knowledge of the complex Plone stack and most of them also don't have knowledge of the Python programming language.

[Nick](https://nickcms.org) is a headless content management system written in Node.js which has an api which is compatible with the Plone restapi. It has a small (~8.800 lines of code) codebase and provides all CMS functionality which Plone provides including the extended permission system.

## Motivation

The motivation to write Nick and use this as a backend is that it uses the same language as used in the current frontend (Volto) which makes it easier for new developers to contribute to the project.

Because Nick is written in Node.js it is asynchronous by default and provides much better [performace](https://www.slideshare.net/slideshow/how-does-nick-compare-to-plone-plone-conference-2024/273734599) (~4x writes and ~100x on bulk operations).

## Assumptions

It is hard for new developers to start working on the Plone backend because of the large stack.

Most frontend developers don't have knowledge of Python.

Maintainance of a large software stack is costly.

## Proposal & Implementation

Our proposal is to replace the Plone backend with [Nick](https://nickcms.org), a backend written in Node.js which is compatible with the Plone rest api.

## Deliverables

- [x] Backend with a rest api compatible with plone.restapi
- [x] Clear migration path for the website data
- [x] Documentation of all the endpoints
- [ ] Documentation and training materials of how to customize the backend

## Risks

None

## Participants

Rob Gietema, @robgietema 
Victor Fernandez de Alba, @sneridagh
Dante Álvarez, @danalvrz

**Comment by stevepiercy** [comment]
More Deliverables to consider:

- [ ] Transfer ownership of the [repository](https://github.com/robgietema/nick) to the Plone GitHub organization.
- [ ] Establish trusted maintainers.
- [ ] Add deployment of the [demo site](https://demo.nickcms.org/) to the @plone/ai-team.
- [ ] Create a [Cookieplone template](https://github.com/plone/cookieplone-templates) `backend-nick`.
- [ ] Decide whether documentation remains standalone, integrated into Plone Documentation, or both.
- [ ] Consider switching documentation to use [Plone Sphinx Theme](https://plone-sphinx-theme.readthedocs.io/).
- [ ] Create a Discord channel `#nick`.

And, of course, discuss with the Steering Circle as soon as possible.

**Comment by tkimnguyen** [comment]
Mind blowing and audacious! Another risk to consider: we will be cast out of the Python orbit… strange to consider.

**Comment by ericof** [comment]
Another important point: The restapi still lacks (or does not care about) most features we have in Plane and Zope, so claiming Nick is compatible with the restapi hides the complexity involved in this endeavor.
We also need to remember that Guillotina claimed, few years back, the same goal of replacing the Phone backend.

**Comment by davisagli** [comment]
I think this is an interesting proposal to keep in mind, but too early to consider seriously. I think it could be considered more seriously at a point in time when multiple Plone providers have independently chosen to use Nick and successfully completed projects with it. (And I don't mean this as stop energy; it might already be a reasonable choice for some projects.)

> Another important point: The restapi still lacks (or does not care about) most features we have in Plone and Zope, so claiming Nick is compatible with the restapi hides the complexity involved in this endeavor.

@ericof Which specific features are you missing?

> We also need to remember that Guillotina claimed, few years back, the same goal of replacing the Plone backend.

This is not quite the same situation, since Nick takes backwards-compatibility more seriously than Guillotina did.

**Comment by petschki** [comment]
"Improve Plone by removing Plone" ... maybe the marketing team wants to think about the wording here?

**Comment by stevepiercy** [comment]
> * [ ]  Documentation and training materials of how to customize the backend

I'm not sure what "customize the backend" means. Does it mean extensibility and add-ons?

Are there any add-ons for Nick?

> too early to consider seriously.

That depends on what one considers Nick to be now and in the future. A _complete replacement_ of the Plone backend is premature now. However it should be considered as an _alternative_ to it now, and a _potential replacement_ in the future, if ever.

As such, I would suggest the following text change from:

> Our proposal is to replace the Plone backend with [Nick](https://nickcms.org/), a backend written in Node.js which is compatible with the Plone rest api.

to:

> Our proposal is to provide an alternative backend for Plone using [Nick](https://nickcms.org/), with full feature-parity to the existing Plone backend.
> Nick is a backend written in Node.js, which is compatible with the Plone [REST API](https://github.com/plone/plone.restapi/).

@petschki @robgietema @davisagli @ericof would this be acceptable?

**Comment by yurj** [comment]
In the Plone backend code there are tenths of thousands of lines, with hundreds of corner cases. How do you think you can replace them? What about advanced Plone features don't covered by REST API? Have you replaced Dexterity and ZCA with few lines of javascript?

**Comment by stevepiercy** [comment]
So far I haven't seen any cited specific features missing from Nick and Volto that currently exist in the Plone backend. I think a distinction must be made between the _implementation of a feature_ and the _feature itself_.

There is mention that there are features in Plone backend that are not implemented in REST API, but I see no specific feature cited.

Dexterity is an implementation of the feature Content Types, which both Nick and Volto have.

ZCA is an implementation of a registry, again which both Nick and Volto have.

IMO, the 1.25M lines of code in the Plone backend is not a feature, but a burden. Since you can do the same things in 8.8K lines of code, it's freedom and liberation.

**Comment by sneridagh** [comment]
Let me leave here some questions.

Would it be so hard to consider this PLIP serious and possible? Who thought that `plone-react`, 7 years later, would have been able to reach the point where Volto has reached? Who would say back then, that finally, we chose and bet on the right technology, the one that won in the industry for dealing with UIs? That we could infuse it with the wisdom and knowledge that we gained through the last decades building CMSs. That we could build a unique and sustainable add-on system, where no other tool in the JS ecosystem has.

I was present ten years ago in the second Bristol Plone Conference open space where Martin Aspeli gave people a reality check stating that rewriting Plone was not possible, when people proposed that we should rewrite Plone in Pyramid.

Well, are we not half way for that to happen? Who says that, if we want to, in another 7 years Nick is not as powerful and feature rich as Plone backend is today? Rob already did it once. The Plone community already did it once. Its name is Volto. Do you really think that we are not capable of doing it again? The point is not in the amount of existing code, but in the wisdom in it and how we transform it on our behalf.

@yurj the Plone community have been implementing Plone projects with Volto since 2017. Which Plone advanced features that REST API does not cover are you missing? Because, this didn't stop us to make business with it, and if it would have, we would have implemented them.

**Comment by sneridagh** [comment]
@mauritsvanrees +1 for rewording "replace" and reformulate the whole story to "complement" and "sum up".

**Comment by ericof** [comment]
First of all, I want to avoid the whole "stop" energy. Sorry if it came across like that.

tl;dr: **I think Nick is no real replacement for Plone anytime soon. Nick could be an option for smaller sites, but larger deployments use features we have been supporting and hardening for more than 25 years.**

# Concerns

So, let's address my specific concerns here

## plone.restapi
> @ericof Which specific features are you missing?

plone.restapi hides the complexity of Plone core from the frontend. Some examples of features of core that are not available:

* Authentication mechanism (PAS is a very good idea, and its complexity also serves many corporate clients with legacy systems and crazy internal rules)
* Content-types framework with reusable behaviors
* Support to multiple workflows to be registered
* Clear add-on contract allowing you to replace parts of Plone with other solutions (PAS, catalog...)

(Does Nick support content rules already?)

Besides that, the restapi, today, does not have (from the top of my mind):

* Constraints support
* Placeful workflow support

## Dropping our entire backend add-on ecosystem

I know most people use only the basics of Plone, but for some of us that need advanced features, the existing add-ons are a great way to be productive, also it is our way of sharing complex features to the community. Moving from Python to Javascript and having to rewrite everything from scratch is not a real option.

## Cost and time involved

To make Nick a viable option would require a major investment in terms of time and resources, which we, as a community, do not have.
If companies (more than one) decide to build projects on top of Nick and invest time there, perfect, otherwise it will be a low-end alternative to what we currently have.

## IP, Trademark, and so on

Some important steps:

* NickCMS needs to be donated to the Plone Foundation: Code and trademarks
* NickCMS codebase needs to be under Plone Foundation / Community control
* The Plone Foundation needs to make sure NickCMS is a trademark we can get domains, social media handles, and other trademarks


# Summary

NickCMS could be a good low-end alternative to the current stack. We could replace GuillotinaCMS from the volto repo with it, but it is not a real replacement for Products.CMFPlone

**Comment by ericof** [comment]
> Our proposal is to provide an alternative backend for Plone using [Nick](https://nickcms.org/), with full feature-parity to the existing Plone backend.

Again, full feature-parity with plone.restapi is not the same of full feature-parity to the existing Plone backend

**Comment by tkimnguyen** [comment]
I think being able to show something even close but with 1% of the source code is pretty !@#$ amazing, and with five years' more development, who knows?

**Comment by robgietema** [comment]
@ericof Thank you for your thorough response. I don't consider this "stop" energy, I am very open to feedback and would like to address some of your concerns:

Authentication mechanism (PAS)
==============================

Nick uses models for handling all specific data like documents, users, roles, permissions etc. These models can be customized in your own project to change the behavior of these models. So implementing a custom authentication mechanism is possible by overriding (parts of) the user, role, group and permission models.

Content-Types framework with reusable behaviors
===============================================

Nick has Content-Types which can be defined using a json-schema or code. They can also have behaviors (and even nested behaviors). So this is pretty similar to what Plone currently provides.

Workflows
=========

You can define as many workflows as you want and the workflows in Nick have the exact same design as used in Plone using states, transitions, permissions etc.

Add-ons and customizabillity
============================

Nick has been build with customizabillity in mind. This means your are able to override almost all of the functionality in Nick like models, routes, events, content types, settings and much more. Nick indeed does not have an ecosystem of backend add-ons yet.

Content rules
=============

Content rules are currently not yet implemented but all building blocks are there like events. This is on my todo and will be released soon(tm).

NickCMS could be a good low-end alternative to the current stack
================================================================

Not sure what low-end means in this context, as listed above almost all functionality is available. If you are referring to the backend eco system then Plone indeed has the edge if you are looking at performance Nick is the high-end system. Nick also has features not available in Plone like async by default, streaming endpoints, (local) llm integration etc.

Again, full feature-parity with plone.restapi is not the same of full feature-parity to the existing Plone backend
==================================================================================================================

When talking about a Volto or Seven setup this is the case, if it is not available in the restapi it is not available to the end user. When looking at Classic that is a different story and this is not about replacing Classic.

To make Nick a viable option would require a major investment in terms of time and resources, which we, as a community, do not have.
====================================================================================================================================

Can you elaborate on this? Are you referring to add-ons or other missing features? Also this does not mean every website running Plone should be upgraded within a week. People can choose to use Nick or stick with the Plone backend the same way we did when Volto was released.

**Comment by yurj** [comment]
@robgietema I think you can add a profile/option for cookieplone, so people can actually start to use it on new projects, and also switch the backend implementation to test it on current projects. That would be a great step forward for Nick adoption. What I still didn't get is how to migrate existing data from Data.fs/Relstorage to Nick and if you took a look at Jensen work on JSONB with Postgresql. I think this in particular can be a good way to move data from the Python implementation to the Nick one.

Aside, Python too can do async and other stuff, and now with 3.14 the GIL lock can be removed. Did you consider Python for Nick implementation or reusing existing JS/Typescript knowledge was the main reason? Thanks for explaining Nick and show that the Plone model is still valid outside the Python implementation.

**Comment by robgietema** [comment]
@yurj thank you for your feedback.

Work is been done on the cookieplone templates as we speak and should be ready soon. In the meanwhile there is an installation guide available for Nick at: https://docs.nickcms.org/usage/installation

Thank you for pointing me to the jsonb implementation that makes the data in the zodb indeed a lot more readable from outside Zope/Plone. The issue with converting the data from there is that not all information is stored in the db (content types for example are stored on disk) and if people don't use relstorage they have to convert to relstorage first. So since Nick uses profiles to setup (and upgrade) your site (also to setup initial content), similar to Plone that is the route I am taking right now. I have created some conversion scripts which take data from plone.exportimport and convert this to profiles. I'm not 100% happy with it and I think the way to go is query the restapi directly and convert from there. Good topic for the next sprint.

I do not know enough about the current Python async implementation but the way I understood it is that you _can_ use Python in Plone/Zope if you want to but by default everything is sync and in Node everything is async by default. Please correct me if I am wrong.

I did not consider Python because indeed the main reason of this project was to use the same language on the frontend and backend so it would be easier for new developers to jump in. This also makes it possible to reuse some of the code (validators etc) between the frontend and backend and even have the option to only use 1 server application which serves both the backend en frontend code using BFF.

I am indeed very happy with the Plone model, the whole system with content types, permissions system, workflows etc is very well thought out and covers all use cases you can imagine. And since the functionality is well described in the plone.restapi docs it was a good reference to start the implementation from.
